=== modified file 'ChangeLog.nih'
--- ChangeLog.nih	2007-02-11 17:01:25 +0000
+++ ChangeLog.nih	2007-02-13 15:42:12 +0000
@@ -1,3 +1,14 @@
+2007-02-13  Scott James Remnant  <scott@netsplit.com>
+
+	* nih/tests/test_option.c (test_parser): gcc doesn't like using
+	variables inside fixed for loops, so initialise it deliberately.
+	* nih/tests/test_command.c (test_help): Was mistakenly repeatedly
+	opening output.
+	(test_parser): Similar for loop issue.
+
+	* nih/signal.c: Update signal name list, some signals don't exist on
+	sparc and some turn up unexpectedly.
+
 2007-02-11  Johan Kiviniemi  <johan@kiviniemi.name>
 
 	* nih/hash.h (NIH_HASH_FOREACH, NIH_HASH_FOREACH_SAFE): Added missing

=== modified file 'nih/signal.c'
--- nih/signal.c	2007-02-06 23:39:10 +0000
+++ nih/signal.c	2007-02-13 15:31:36 +0000
@@ -71,6 +71,9 @@
 	{ SIGTRAP,   "TRAP"   },
 	{ SIGABRT,   "ABRT"   },
 	{ SIGIOT,    "IOT"    },
+#ifdef SIGEMT
+	{ SIGEMT,    "EMT"    },
+#endif
 	{ SIGBUS,    "BUS"    },
 	{ SIGFPE,    "FPE"    },
 	{ SIGKILL,   "KILL"   },
@@ -80,12 +83,14 @@
 	{ SIGPIPE,   "PIPE"   },
 	{ SIGALRM,   "ALRM"   },
 	{ SIGTERM,   "TERM"   },
+#ifdef SIGSTKFLT
 	{ SIGSTKFLT, "STKFLT" },
+#endif
 	{ SIGCHLD,   "CHLD"   },
 	{ SIGCLD,    "CLD"    },
 	{ SIGCONT,   "CONT"   },
 	{ SIGSTOP,   "STOP"   },
-	{ SIGTSTP,   "TSTP"  },
+	{ SIGTSTP,   "TSTP"   },
 	{ SIGTTIN,   "TTIN"   },
 	{ SIGTTOU,   "TTOU"   },
 	{ SIGURG,    "URG"    },
@@ -96,6 +101,9 @@
 	{ SIGWINCH,  "WINCH"  },
 	{ SIGIO,     "IO"     },
 	{ SIGPOLL,   "POLL"   },
+#ifdef SIGLOST
+	{ SIGLOST,   "LOST"   },
+#endif
 	{ SIGPWR,    "PWR"    },
 	{ SIGSYS,    "SYS"    },
 	{ SIGUNUSED, "UNUSED" },

=== modified file 'nih/tests/test_command.c'
--- nih/tests/test_command.c	2007-01-22 02:07:49 +0000
+++ nih/tests/test_command.c	2007-02-13 15:42:09 +0000
@@ -137,7 +137,7 @@
 {
 	FILE *output;
 	char *argv[16];
-	int   argc, ret;
+	int   argc, ret = 0;
 
 	TEST_FUNCTION ("nih_command_parser");
 	program_name = "test";
@@ -593,6 +593,7 @@
 	int    argc, status;
 
 	TEST_FUNCTION ("nih_command_help");
+	output = tmpfile ();
 
 	/* Check that we can obtain a list of command using the "help"
 	 * command; which terminates the process with exit code 0.  The
@@ -608,7 +609,6 @@
 		argv[argc++] = "help";
 		argv[argc] = NULL;
 
-		output = tmpfile ();
 		TEST_CHILD (pid) {
 			unsetenv ("COLUMNS");
 
@@ -664,7 +664,6 @@
 		argv[argc++] = "help";
 		argv[argc] = NULL;
 
-		output = tmpfile ();
 		TEST_CHILD (pid) {
 			unsetenv ("COLUMNS");
 

=== modified file 'nih/tests/test_option.c'
--- nih/tests/test_option.c	2007-02-03 15:35:38 +0000
+++ nih/tests/test_option.c	2007-02-13 15:40:30 +0000
@@ -95,7 +95,7 @@
 test_parser (void)
 {
 	FILE *output;
-	char *argv[16], **args;
+	char *argv[16], **args = NULL;
 	int   argc;
 
 	TEST_FUNCTION ("nih_option_parser");

