Index: upstart/nih/Makefile.am
===================================================================
--- upstart.orig/nih/Makefile.am	2007-10-03 13:00:21.000000000 +0200
+++ upstart/nih/Makefile.am	2009-03-12 04:10:12.000000000 +0100
@@ -29,7 +29,7 @@
 
 
 libnih_la_LDFLAGS = \
-	-version-info 0:0:0
+	-static -version-info 0:0:0
 if HAVE_VERSION_SCRIPT_ARG
 libnih_la_LDFLAGS += @VERSION_SCRIPT_ARG@=$(srcdir)/libnih.ver
 endif
Index: upstart/nih/Makefile.in
===================================================================
--- upstart.orig/nih/Makefile.in	2007-10-11 23:12:44.000000000 +0200
+++ upstart/nih/Makefile.in	2009-03-12 04:10:12.000000000 +0100
@@ -403,7 +403,7 @@
 	logging.c \
 	error.c
 
-libnih_la_LDFLAGS = -version-info 0:0:0 $(am__append_1)
+libnih_la_LDFLAGS = -static -version-info 0:0:0 $(am__append_1)
 EXTRA_DIST = libnih.ver libnih.supp
 include_HEADERS = \
 	libnih.h
Index: upstart/upstart/Makefile.am
===================================================================
--- upstart.orig/upstart/Makefile.am	2007-10-11 22:59:12.000000000 +0200
+++ upstart/upstart/Makefile.am	2009-03-12 04:10:12.000000000 +0100
@@ -15,7 +15,7 @@
 
 
 libupstart_la_LDFLAGS = \
-	-version-info 0:0:0
+	-static -version-info 0:0:0
 if HAVE_VERSION_SCRIPT_ARG
 libupstart_la_LDFLAGS += @VERSION_SCRIPT_ARG@=$(srcdir)/libupstart.ver
 endif
Index: upstart/upstart/Makefile.in
===================================================================
--- upstart.orig/upstart/Makefile.in	2007-10-11 23:12:44.000000000 +0200
+++ upstart/upstart/Makefile.in	2009-03-12 04:10:12.000000000 +0100
@@ -281,7 +281,7 @@
 	message.c \
 	wire.c
 
-libupstart_la_LDFLAGS = -version-info 0:0:0 $(am__append_1)
+libupstart_la_LDFLAGS = -static -version-info 0:0:0 $(am__append_1)
 libupstart_la_LIBADD = \
 	../nih/libnih.la \
 	$(LTLIBINTL)
