Index: upstart/Makefile.in
===================================================================
--- upstart.orig/Makefile.in	2010-06-18 06:26:59.515931387 +0200
+++ upstart/Makefile.in	2010-06-18 06:27:06.319928496 +0200
@@ -232,6 +232,8 @@
 PRI_MACROS_BROKEN = @PRI_MACROS_BROKEN@
 RANLIB = @RANLIB@
 SED = @SED@
+SELINUX_CFLAGS = @SELINUX_CFLAGS@
+SELINUX_LIBS = @SELINUX_LIBS@
 SET_MAKE = @SET_MAKE@
 SHELL = @SHELL@
 STRIP = @STRIP@
Index: upstart/conf/Makefile.in
===================================================================
--- upstart.orig/conf/Makefile.in	2010-06-18 06:26:59.567930752 +0200
+++ upstart/conf/Makefile.in	2010-06-18 06:27:06.323929565 +0200
@@ -200,6 +200,8 @@
 PRI_MACROS_BROKEN = @PRI_MACROS_BROKEN@
 RANLIB = @RANLIB@
 SED = @SED@
+SELINUX_CFLAGS = @SELINUX_CFLAGS@
+SELINUX_LIBS = @SELINUX_LIBS@
 SET_MAKE = @SET_MAKE@
 SHELL = @SHELL@
 STRIP = @STRIP@
Index: upstart/config.h.in
===================================================================
--- upstart.orig/config.h.in	2010-06-18 06:26:59.543929926 +0200
+++ upstart/config.h.in	2010-06-18 06:27:06.323929565 +0200
@@ -152,6 +152,9 @@
 /* Define to 1 if you have the `putenv' function. */
 #undef HAVE_PUTENV
 
+/* Define if we have SELinux */
+#undef HAVE_SELINUX
+
 /* Define to 1 if you have the `setenv' function. */
 #undef HAVE_SETENV
 
Index: upstart/configure
===================================================================
--- upstart.orig/configure	2010-06-18 06:26:59.591931019 +0200
+++ upstart/configure	2010-06-18 06:27:06.335934726 +0200
@@ -751,6 +751,8 @@
 am__EXEEXT_TRUE
 LTLIBOBJS
 LIBOBJS
+SELINUX_LIBS
+SELINUX_CFLAGS
 DBUS_LIBS
 DBUS_CFLAGS
 NIH_DBUS_LIBS
@@ -937,6 +939,7 @@
 with_included_gettext
 with_libintl_prefix
 with_local_libnih
+enable_selinux
 enable_threading
 enable_compiler_warnings
 enable_compiler_optimisations
@@ -958,7 +961,9 @@
 NIH_DBUS_CFLAGS
 NIH_DBUS_LIBS
 DBUS_CFLAGS
-DBUS_LIBS'
+DBUS_LIBS
+SELINUX_CFLAGS
+SELINUX_LIBS'
 
 
 # Initialize some variables set by options.
@@ -1594,6 +1599,7 @@
                           specify multithreading API
   --disable-threads       build without multithread safety
   --disable-rpath         do not hardcode runtime library paths
+  --enable-selinux        enable SELinux support
   --enable-threading      Enable support for multi-threading
   --enable-compiler-warnings
                           Enable additional compiler warnings
@@ -1639,6 +1645,10 @@
               linker flags for NIH_DBUS, overriding pkg-config
   DBUS_CFLAGS C compiler flags for DBUS, overriding pkg-config
   DBUS_LIBS   linker flags for DBUS, overriding pkg-config
+  SELINUX_CFLAGS
+              C compiler flags for SELINUX, overriding pkg-config
+  SELINUX_LIBS
+              linker flags for SELINUX, overriding pkg-config
 
 Use these variables to override the choices made by `configure' or to help
 it to find libraries and programs with nonstandard names/locations.
@@ -5215,13 +5225,13 @@
 else
   lt_cv_nm_interface="BSD nm"
   echo "int some_variable = 0;" > conftest.$ac_ext
-  (eval echo "\"\$as_me:5218: $ac_compile\"" >&5)
+  (eval echo "\"\$as_me:5228: $ac_compile\"" >&5)
   (eval "$ac_compile" 2>conftest.err)
   cat conftest.err >&5
-  (eval echo "\"\$as_me:5221: $NM \\\"conftest.$ac_objext\\\"\"" >&5)
+  (eval echo "\"\$as_me:5231: $NM \\\"conftest.$ac_objext\\\"\"" >&5)
   (eval "$NM \"conftest.$ac_objext\"" 2>conftest.err > conftest.out)
   cat conftest.err >&5
-  (eval echo "\"\$as_me:5224: output\"" >&5)
+  (eval echo "\"\$as_me:5234: output\"" >&5)
   cat conftest.out >&5
   if $GREP 'External.*some_variable' conftest.out > /dev/null; then
     lt_cv_nm_interface="MS dumpbin"
@@ -6426,7 +6436,7 @@
   ;;
 *-*-irix6*)
   # Find out which ABI we are using.
-  echo '#line 6429 "configure"' > conftest.$ac_ext
+  echo '#line 6439 "configure"' > conftest.$ac_ext
   if { { eval echo "\"\$as_me\":${as_lineno-$LINENO}: \"$ac_compile\""; } >&5
   (eval $ac_compile) 2>&5
   ac_status=$?
@@ -7688,11 +7698,11 @@
    -e 's:.*FLAGS}\{0,1\} :&$lt_compiler_flag :; t' \
    -e 's: [^ ]*conftest\.: $lt_compiler_flag&:; t' \
    -e 's:$: $lt_compiler_flag:'`
-   (eval echo "\"\$as_me:7691: $lt_compile\"" >&5)
+   (eval echo "\"\$as_me:7701: $lt_compile\"" >&5)
    (eval "$lt_compile" 2>conftest.err)
    ac_status=$?
    cat conftest.err >&5
-   echo "$as_me:7695: \$? = $ac_status" >&5
+   echo "$as_me:7705: \$? = $ac_status" >&5
    if (exit $ac_status) && test -s "$ac_outfile"; then
      # The compiler can only warn and ignore the option if not recognized
      # So say no if there are warnings other than the usual output.
@@ -8027,11 +8037,11 @@
    -e 's:.*FLAGS}\{0,1\} :&$lt_compiler_flag :; t' \
    -e 's: [^ ]*conftest\.: $lt_compiler_flag&:; t' \
    -e 's:$: $lt_compiler_flag:'`
-   (eval echo "\"\$as_me:8030: $lt_compile\"" >&5)
+   (eval echo "\"\$as_me:8040: $lt_compile\"" >&5)
    (eval "$lt_compile" 2>conftest.err)
    ac_status=$?
    cat conftest.err >&5
-   echo "$as_me:8034: \$? = $ac_status" >&5
+   echo "$as_me:8044: \$? = $ac_status" >&5
    if (exit $ac_status) && test -s "$ac_outfile"; then
      # The compiler can only warn and ignore the option if not recognized
      # So say no if there are warnings other than the usual output.
@@ -8132,11 +8142,11 @@
    -e 's:.*FLAGS}\{0,1\} :&$lt_compiler_flag :; t' \
    -e 's: [^ ]*conftest\.: $lt_compiler_flag&:; t' \
    -e 's:$: $lt_compiler_flag:'`
-   (eval echo "\"\$as_me:8135: $lt_compile\"" >&5)
+   (eval echo "\"\$as_me:8145: $lt_compile\"" >&5)
    (eval "$lt_compile" 2>out/conftest.err)
    ac_status=$?
    cat out/conftest.err >&5
-   echo "$as_me:8139: \$? = $ac_status" >&5
+   echo "$as_me:8149: \$? = $ac_status" >&5
    if (exit $ac_status) && test -s out/conftest2.$ac_objext
    then
      # The compiler can only warn and ignore the option if not recognized
@@ -8187,11 +8197,11 @@
    -e 's:.*FLAGS}\{0,1\} :&$lt_compiler_flag :; t' \
    -e 's: [^ ]*conftest\.: $lt_compiler_flag&:; t' \
    -e 's:$: $lt_compiler_flag:'`
-   (eval echo "\"\$as_me:8190: $lt_compile\"" >&5)
+   (eval echo "\"\$as_me:8200: $lt_compile\"" >&5)
    (eval "$lt_compile" 2>out/conftest.err)
    ac_status=$?
    cat out/conftest.err >&5
-   echo "$as_me:8194: \$? = $ac_status" >&5
+   echo "$as_me:8204: \$? = $ac_status" >&5
    if (exit $ac_status) && test -s out/conftest2.$ac_objext
    then
      # The compiler can only warn and ignore the option if not recognized
@@ -10571,7 +10581,7 @@
   lt_dlunknown=0; lt_dlno_uscore=1; lt_dlneed_uscore=2
   lt_status=$lt_dlunknown
   cat > conftest.$ac_ext <<_LT_EOF
-#line 10574 "configure"
+#line 10584 "configure"
 #include "confdefs.h"
 
 #if HAVE_DLFCN_H
@@ -10667,7 +10677,7 @@
   lt_dlunknown=0; lt_dlno_uscore=1; lt_dlneed_uscore=2
   lt_status=$lt_dlunknown
   cat > conftest.$ac_ext <<_LT_EOF
-#line 10670 "configure"
+#line 10680 "configure"
 #include "confdefs.h"
 
 #if HAVE_DLFCN_H
@@ -16842,6 +16852,110 @@
 	:
 fi
 
+# Check whether --enable-selinux was given.
+if test "${enable_selinux+set}" = set; then :
+  enableval=$enable_selinux;
+else
+  enable_selinux=no
+fi
+
+
+if test "x$enable_selinux" = "xyes" ; then
+
+pkg_failed=no
+{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for SELINUX" >&5
+$as_echo_n "checking for SELINUX... " >&6; }
+
+if test -n "$PKG_CONFIG"; then
+    if test -n "$SELINUX_CFLAGS"; then
+        pkg_cv_SELINUX_CFLAGS="$SELINUX_CFLAGS"
+    else
+        if test -n "$PKG_CONFIG" && \
+    { { $as_echo "$as_me:${as_lineno-$LINENO}: \$PKG_CONFIG --exists --print-errors \"libselinux\""; } >&5
+  ($PKG_CONFIG --exists --print-errors "libselinux") 2>&5
+  ac_status=$?
+  $as_echo "$as_me:${as_lineno-$LINENO}: \$? = $ac_status" >&5
+  test $ac_status = 0; }; then
+  pkg_cv_SELINUX_CFLAGS=`$PKG_CONFIG --cflags "libselinux" 2>/dev/null`
+else
+  pkg_failed=yes
+fi
+    fi
+else
+	pkg_failed=untried
+fi
+if test -n "$PKG_CONFIG"; then
+    if test -n "$SELINUX_LIBS"; then
+        pkg_cv_SELINUX_LIBS="$SELINUX_LIBS"
+    else
+        if test -n "$PKG_CONFIG" && \
+    { { $as_echo "$as_me:${as_lineno-$LINENO}: \$PKG_CONFIG --exists --print-errors \"libselinux\""; } >&5
+  ($PKG_CONFIG --exists --print-errors "libselinux") 2>&5
+  ac_status=$?
+  $as_echo "$as_me:${as_lineno-$LINENO}: \$? = $ac_status" >&5
+  test $ac_status = 0; }; then
+  pkg_cv_SELINUX_LIBS=`$PKG_CONFIG --libs "libselinux" 2>/dev/null`
+else
+  pkg_failed=yes
+fi
+    fi
+else
+	pkg_failed=untried
+fi
+
+
+
+if test $pkg_failed = yes; then
+
+if $PKG_CONFIG --atleast-pkgconfig-version 0.20; then
+        _pkg_short_errors_supported=yes
+else
+        _pkg_short_errors_supported=no
+fi
+        if test $_pkg_short_errors_supported = yes; then
+	        SELINUX_PKG_ERRORS=`$PKG_CONFIG --short-errors --errors-to-stdout --print-errors "libselinux"`
+        else
+	        SELINUX_PKG_ERRORS=`$PKG_CONFIG --errors-to-stdout --print-errors "libselinux"`
+        fi
+	# Put the nasty error message in config.log where it belongs
+	echo "$SELINUX_PKG_ERRORS" >&5
+
+	as_fn_error "Package requirements (libselinux) were not met:
+
+$SELINUX_PKG_ERRORS
+
+Consider adjusting the PKG_CONFIG_PATH environment variable if you
+installed software in a non-standard prefix.
+
+Alternatively, you may set the environment variables SELINUX_CFLAGS
+and SELINUX_LIBS to avoid the need to call pkg-config.
+See the pkg-config man page for more details.
+" "$LINENO" 5
+elif test $pkg_failed = untried; then
+	{ { $as_echo "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
+$as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
+as_fn_error "The pkg-config script could not be found or is too old.  Make sure it
+is in your PATH or set the PKG_CONFIG environment variable to the full
+path to pkg-config.
+
+Alternatively, you may set the environment variables SELINUX_CFLAGS
+and SELINUX_LIBS to avoid the need to call pkg-config.
+See the pkg-config man page for more details.
+
+To get pkg-config, see <http://pkg-config.freedesktop.org/>.
+See \`config.log' for more details." "$LINENO" 5; }
+else
+	SELINUX_CFLAGS=$pkg_cv_SELINUX_CFLAGS
+	SELINUX_LIBS=$pkg_cv_SELINUX_LIBS
+        { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes" >&5
+$as_echo "yes" >&6; }
+	:
+fi
+
+$as_echo "#define HAVE_SELINUX 1" >>confdefs.h
+
+fi
+
 # Checks for header files.
 for ac_header in valgrind/valgrind.h
 do :
Index: upstart/contrib/Makefile.in
===================================================================
--- upstart.orig/contrib/Makefile.in	2010-06-18 06:26:59.491932241 +0200
+++ upstart/contrib/Makefile.in	2010-06-18 06:27:06.343929878 +0200
@@ -175,6 +175,8 @@
 PRI_MACROS_BROKEN = @PRI_MACROS_BROKEN@
 RANLIB = @RANLIB@
 SED = @SED@
+SELINUX_CFLAGS = @SELINUX_CFLAGS@
+SELINUX_LIBS = @SELINUX_LIBS@
 SET_MAKE = @SET_MAKE@
 SHELL = @SHELL@
 STRIP = @STRIP@
Index: upstart/dbus/Makefile.in
===================================================================
--- upstart.orig/dbus/Makefile.in	2010-06-18 06:26:59.503930138 +0200
+++ upstart/dbus/Makefile.in	2010-06-18 06:27:06.347924801 +0200
@@ -200,6 +200,8 @@
 PRI_MACROS_BROKEN = @PRI_MACROS_BROKEN@
 RANLIB = @RANLIB@
 SED = @SED@
+SELINUX_CFLAGS = @SELINUX_CFLAGS@
+SELINUX_LIBS = @SELINUX_LIBS@
 SET_MAKE = @SET_MAKE@
 SHELL = @SHELL@
 STRIP = @STRIP@
Index: upstart/doc/Makefile.in
===================================================================
--- upstart.orig/doc/Makefile.in	2010-06-18 06:26:59.579930886 +0200
+++ upstart/doc/Makefile.in	2010-06-18 06:27:06.347924801 +0200
@@ -175,6 +175,8 @@
 PRI_MACROS_BROKEN = @PRI_MACROS_BROKEN@
 RANLIB = @RANLIB@
 SED = @SED@
+SELINUX_CFLAGS = @SELINUX_CFLAGS@
+SELINUX_LIBS = @SELINUX_LIBS@
 SET_MAKE = @SET_MAKE@
 SHELL = @SHELL@
 STRIP = @STRIP@
Index: upstart/init/Makefile.in
===================================================================
--- upstart.orig/init/Makefile.in	2010-06-18 06:26:59.555928663 +0200
+++ upstart/init/Makefile.in	2010-06-18 06:27:06.347924801 +0200
@@ -95,7 +95,8 @@
 init_OBJECTS = $(am_init_OBJECTS) $(nodist_init_OBJECTS)
 am__DEPENDENCIES_1 =
 init_DEPENDENCIES = $(am__DEPENDENCIES_1) $(am__DEPENDENCIES_1) \
-	$(am__DEPENDENCIES_1) $(am__DEPENDENCIES_1)
+	$(am__DEPENDENCIES_1) $(am__DEPENDENCIES_1) \
+	$(am__DEPENDENCIES_1)
 AM_V_lt = $(am__v_lt_$(V))
 am__v_lt_ = $(am__v_lt_$(AM_DEFAULT_VERBOSITY))
 am__v_lt_0 = --silent
@@ -383,6 +384,8 @@
 PRI_MACROS_BROKEN = @PRI_MACROS_BROKEN@
 RANLIB = @RANLIB@
 SED = @SED@
+SELINUX_CFLAGS = @SELINUX_CFLAGS@
+SELINUX_LIBS = @SELINUX_LIBS@
 SET_MAKE = @SET_MAKE@
 SHELL = @SHELL@
 STRIP = @STRIP@
@@ -451,7 +454,8 @@
 AM_CFLAGS = \
 	$(NIH_CFLAGS) \
 	$(NIH_DBUS_CFLAGS) \
-	$(DBUS_CFLAGS)
+	$(DBUS_CFLAGS) \
+	$(SELINUX_CFLAGS)
 
 AM_CPPFLAGS = \
 	-DLOCALEDIR="\"$(localedir)\"" \
@@ -504,6 +508,7 @@
 	$(NIH_LIBS) \
 	$(NIH_DBUS_LIBS) \
 	$(DBUS_LIBS) \
+	$(SELINUX_LIBS) \
 	-lrt
 
 com_ubuntu_Upstart_OUTPUTS = \
Index: upstart/util/Makefile.in
===================================================================
--- upstart.orig/util/Makefile.in	2010-06-18 06:26:59.531930072 +0200
+++ upstart/util/Makefile.in	2010-06-18 06:27:06.351929780 +0200
@@ -316,6 +316,8 @@
 PRI_MACROS_BROKEN = @PRI_MACROS_BROKEN@
 RANLIB = @RANLIB@
 SED = @SED@
+SELINUX_CFLAGS = @SELINUX_CFLAGS@
+SELINUX_LIBS = @SELINUX_LIBS@
 SET_MAKE = @SET_MAKE@
 SHELL = @SHELL@
 STRIP = @STRIP@
